### PR TITLE
Change how u8 slices/array are formatted with {}

### PR DESCRIPTION
**Pull request by LemonBoy**
This PR spiraled out of control and became a monster that removes the special-casing of `[]u8/[N]u8/*[N]u8` as strings.

Now `{}` defaults to printing the array or slice elements as numbers, If you _want_ a string you _need_ `{s}`.

### CODE REVIEW

File: lib/std/fmt.zig [hunk 1]
```diff
@@ -453,13 +480,18 @@ pub fn formatType(
...
-                    return formatText(value, fmt, options, writer);
-                }
                 if (ptr_info.child == u8) {
                     return formatText(value, fmt, options, writer);
                 }
-                return format(writer, "{}@{x}", .{ @typeName(ptr_info.child), @ptrToInt(value.ptr) });
+
+                try writer.writeAll("[");
+                for (value) |elem, i| {
+                    try formatType(elem, fmt, options, writer, max_depth);
```
Comment (data-man) [inline lib/std/fmt.zig:489]:
```suggestion
                    try formatType(elem, fmt, options, writer, max_depth - 1);
```

File: lib/std/fmt.zig [hunk 2]
```diff
@@ -453,13 +480,18 @@ pub fn formatType(
                 return format(writer, "{}@{x}", .{ @typeName(@typeInfo(T).Pointer.child), @ptrToInt(value) });
             },
             .Slice => {
```
Comment (data-man) [inline lib/std/fmt.zig:482]:
```suggestion
            .Slice => {
                if (max_depth == 0) {
                    return writer.writeAll("[ ... ]");
                }
```

File: lib/std/fmt.zig [hunk 3]
```diff
@@ -453,13 +480,21 @@ pub fn formatType(
...
-                if (fmt.len > 0 and ((fmt[0] == 'x') or (fmt[0] == 'X'))) {
-                    return formatText(value, fmt, options, writer);
+                if (max_depth == 0) {
+                    return writer.writeAll("[ ... ]");
                 }
                 if (ptr_info.child == u8) {
```
[the remark this answers is not in the document]
Comment (LemonBoy) [reply lib/std/fmt.zig:486]: Wait, then `formatText` cannot handle `x` nor `e`...

[the remark this answers is not in the document]
Comment (data-man) [reply lib/std/fmt.zig:486]: `if (ptr_info.child == u8 and fmt.len == 0 or (fmt.len > 0 and (fmt[0] == 'x' or fmt[0] == 'X')))`
?

**Comment by LemonBoy** [comment]
We should really sort this mess with all the formatting specifiers & their defaults :(

**Comment by data-man** [comment]
Why so few thumbs up? :smile:

**Comment by data-man** [comment]
Dear maintainers, is there a chance this PR will be included in the upcoming release?

**Comment by jedisct1** [comment]
+1. That would be very useful.

**Comment by LemonBoy** [comment]
The FreeBSD CI is acting up again, cool.

> Dear maintainers, is there a chance this PR will be included in the upcoming release?

The different handling of `u8` arrays/slices is still up for discussion, let's take our time to discuss about every detail before slamming the merge button.

File: lib/std/fmt.zig [hunk 4]
```diff
@@ -434,55 +460,67 @@ pub fn formatType(
...
-                else => return format(writer, "{}@{x}", .{ @typeName(@typeInfo(T).Pointer.child), @ptrToInt(value) }),
+                else => return format(writer, "{}@{x}", .{ @typeName(ptr_info.child), @ptrToInt(value) }),
             },
             .Many, .C => {
                 if (ptr_info.sentinel) |sentinel| {
                     return formatType(mem.span(value), fmt, options, writer, max_depth);
                 }
                 if (ptr_info.child == u8) {
-                    if (fmt.len > 0 and fmt[0] == 's') {
+                    if (fmt.len > 0 and comptime mem.indexOfScalar(u8, "sxXzZ", fmt[0]) != null) {
```
Comment (Vexu) [inline lib/std/fmt.zig:475]: `formatText` also takes `eE`, though that should probably be merged with `Z` which does the same but using Zig escape sequences.

File: lib/std/fmt.zig [hunk 5]
```diff
@@ -1381,11 +1420,20 @@ test "slice" {
...
+        try testFmt("slice: []const u8@deadbeef\n", "slice: {*}\n", .{value});
     }
 
     try testFmt("buf:  Test\n", "buf: {s:5}\n", .{"Test"});
     try testFmt("buf: Test\n Other text", "buf: {s}\n Other text", .{"Test"});
+
+    {
+        var int_slice = [_]u32{ 1, 4096, 391891, 1111111111 };
+        var runtime_zero: usize = 0;
+        try testFmt("int: [1, 4096, 391891, 1111111111]", "int: {}", .{int_slice[runtime_zero..]});
```
Comment (Vexu) [inline lib/std/fmt.zig:1432]: Is there a reason why slices don't use curly braces like arrays?

Comment (LemonBoy) [reply lib/std/fmt.zig:1432]: Yeah, style-wise there are a few questions. Do we want a space after the opening/closing brace? Do we want curly or square brackets? Same symbol for both arrays and slices?

Comment (Vexu) [reply lib/std/fmt.zig:1432]: If we want to make generating Zig with the formatting system easier we should definitely use curly braces for both. For the spaces I'd either follow `zig fmt` or just leave it as is to keep things simple.

**Comment by LemonBoy** [comment]
If we're in for a bold breaking (and perhaps controversial) change we could:
* Make the default (`{}`) specifier for arrays print the content as if their elements were numbers
* Strings require an explicit `{s}` (or `{z/Z}`) as there's no separate string type

This way we can get rid of the special-case handling of `[]u8/[N]u8` as strings and streamline a bit the formatting rules.

**Comment by MasterQ32** [comment]
Yes, please.  We need at least three options then:
- `{}` print as array
- `{s}` print as raw string
- `{q}` print as quoted string ("\xF3" will yield "\\xF3" instead of the raw byte output)

**Comment by andrewrk** [comment]
> If we're in for a bold breaking (and perhaps controversial) change we could:

I support this. I think `{}` in format strings is already a footgun because it can be a runtime bug if you expected that your parameter was a string-like object but instead it is a struct or something and ends up printing a struct representation of it.

**Comment by LemonBoy** [comment]
> Yes, please. We need at least three options then:

Right now the supported specifiers are:
* `{}` to print the value as an array,
* `{s}` to print the value as a string
* `{z}`/`{Z}` to print the value as a valid Zig identifier
* `{x}`/`{X}` to print the content in hexadecimal form (eg. `[0x12, 0x34]` -> "1234")
  - This conflicts with `xX` for the array values, should find a new name for this.
  - Do we need it? Does it deserve its own specifier? Can we spin it into its own `hexdump`-like function if it's really needed?
* `{e}/`{E}` to print the content as an ASCII string, the non-printable characters are escaped (`\xXX`)
  - Limited to ASCII strings :( 
  - People never find it! Maybe because it uses the same specifier used for printing floating point numbers...
  - Is this what people want from a `{q}` (for quote) specifier?

The last two specifiers should be rehomed/renamed to complete the plan.

**Comment by MasterQ32** [comment]
> Do we need it? Does it deserve its own specifier? Can we spin it into its own hexdump-like function if it's really needed?

Yes, very. I use it often to verify that binary data actually looks like it should and i also use it for several hexdump style outputs

> Is this what people want from a {q} (for quote) specifier?

We could introduce `{q}` for general string quoting (everything non-valid utf-8 is quoted) and use `{a}` for ASCII quoted strings

**Comment by LemonBoy** [comment]
> Yes, very. I use it often to verify that binary data actually looks like it should and i also use it for several hexdump style outputs

Fair enough, where do we put it? `{sx}` maybe?

> We could introduce {q} for general string quoting (everything non-valid utf-8 is quoted) and use {a} for ASCII quoted strings

Invalid UTF-8 should be replaced with `U+FFFD` according to the sacred texts, what I had in mind was to traverse the string in terms of utf8 runes and apply the usual `ascii.isprint` predicate on them. It won't escape unprintable non-ASCII codepoints but that's IMO better than splitting the specifier in two.
I do agree that `{q}` is a better name than `{e}`.

File: src/Compilation.zig [hunk 6]
```diff
@@ -2361,15 +2361,15 @@ pub fn generateBuiltinZigSource(comp: *Compilation, allocator: *Allocator) ![]u8
         \\pub const arch = Target.current.cpu.arch;
         \\/// Deprecated
         \\pub const endian = Target.current.cpu.arch.endian();
-        \\pub const output_mode = OutputMode.{};
-        \\pub const link_mode = LinkMode.{};
+        \\pub const output_mode = OutputMode.{s};
+        \\pub const link_mode = LinkMode.{s};
```
Comment (Vexu) [inline src/Compilation.zig:2365]: All the `{s}` in `generateBuiltinZigSource` should be `{z}`.

**Comment by marler8997** [comment]
I built commit 061b42738097f08915a3458c4014a8f146d7731e, I was able to produce a memory leak error running the command below:

```
D:\git\zig\build>.\bin\zig.exe test D:\git\zig\lib\std\std.zig --test-name-prefix std-native-Debug-bare-multi  --cache-dir D:\git\zig\zig-cache --name test -I D:\git\zig\test --override-lib-dir D:\git\zig\lib
Test [205/1614] os.test.test "std-native-Debug-bare-multisymlink with relative paths"...SKIP
Test [207/1614] os.test.test "std-native-Debug-bare-multifstatat"...SKIP
Test [208/1614] os.test.test "std-native-Debug-bare-multireadlinkat"...SKIP
Test [215/1614] os.test.test "std-native-Debug-bare-multisigaltstack"...SKIP
Test [216/1614] os.test.test "std-native-Debug-bare-multidl_iterate_phdr"...SKIP
Test [217/1614] os.test.test "std-native-Debug-bare-multigethostname"...SKIP
Test [218/1614] os.test.test "std-native-Debug-bare-multipipe"...SKIP
Test [220/1614] os.test.test "std-native-Debug-bare-multimemfd_create"...SKIP
Test [221/1614] os.test.test "std-native-Debug-bare-multimmap"...SKIP
Test [223/1614] os.test.test "std-native-Debug-bare-multifcntl"...SKIP
Test [224/1614] os.test.test "std-native-Debug-bare-multisignalfd"...SKIP
Test [225/1614] os.test.test "std-native-Debug-bare-multisync"...SKIP
Test [227/1614] os.test.test "std-native-Debug-bare-multigetrlimit and setrlimit"...SKIP
Test [325/1614] packed_int_array.test "std-native-Debug-bare-multiPackedIntArray"...SKIP
Test [326/1614] packed_int_array.test "std-native-Debug-bare-multiPackedIntArray init"...SKIP
Test [327/1614] packed_int_array.test "std-native-Debug-bare-multiPackedIntSlice"...SKIP
Test [328/1614] packed_int_array.test "std-native-Debug-bare-multiPackedIntSlice of PackedInt(Array/Slice)"...SKIP
Test [329/1614] packed_int_array.test "std-native-Debug-bare-multiPackedIntSlice accumulating bit offsets"...SKIP
Test [330/1614] packed_int_array.test "std-native-Debug-bare-multiPackedInt(Array/Slice) sliceCast"...SKIP
Test [331/1614] packed_int_array.test "std-native-Debug-bare-multiPackedInt(Array/Slice)Endian"...SKIP
Test [332/1614] packed_int_array.test "std-native-Debug-bare-multiPackedIntArray at end of available memory"...SKIP
Test [333/1614] packed_int_array.test "std-native-Debug-bare-multiPackedIntSlice at end of available memory"...SKIP
Test [351/1614] progress.test "std-native-Debug-bare-multibasic functionality"...SKIP
Test [423/1614] process.test "std-native-Debug-bare-multios.getEnvMap"... [std] (err): Memory leak detected:


[std] (err): Memory leak detected:


[std] (err): Memory leak detected:


Test [478/1614] event.loop.test "std-native-Debug-bare-multistd.event.Loop - basic"...SKIP
Test [479/1614] event.loop.test "std-native-Debug-bare-multistd.event.Loop - runDetached"...SKIP
Test [480/1614] event.loop.test "std-native-Debug-bare-multistd.event.Loop - sleep"...SKIP
Test [681/1614] os.linux.io_uring.test "std-native-Debug-bare-multistructs/offsets/entries"...SKIP
Test [682/1614] os.linux.io_uring.test "std-native-Debug-bare-multinop"...SKIP
Test [683/1614] os.linux.io_uring.test "std-native-Debug-bare-multireadv"...SKIP
Test [684/1614] os.linux.io_uring.test "std-native-Debug-bare-multiwritev/fsync/readv"...SKIP
Test [685/1614] os.linux.io_uring.test "std-native-Debug-bare-multiwrite/read"...SKIP
Test [686/1614] os.linux.io_uring.test "std-native-Debug-bare-multiopenat"...SKIP
Test [687/1614] os.linux.io_uring.test "std-native-Debug-bare-multiclose"...SKIP
Test [688/1614] os.linux.io_uring.test "std-native-Debug-bare-multiaccept/connect/send/recv"...SKIP
Test [694/1614] io.c_writer.test "std-native-Debug-bare-multi"...SKIP
Test [705/1614] fs.test.test "std-native-Debug-bare-multiDir.readLink"...SKIP
Test [708/1614] fs.test.test "std-native-Debug-bare-multireadLinkAbsolute"...SKIP
Test [716/1614] fs.test.test "std-native-Debug-bare-multiDir.rename directories"...SKIP
Test [717/1614] fs.test.test "std-native-Debug-bare-multiDir.rename file <-> dir"...SKIP
Test [731/1614] fs.test.test "std-native-Debug-bare-multiopen file with exclusive lock twice, make sure it waits"...SKIP
Test [734/1614] fs.watch.test "std-native-Debug-bare-multiwrite a file, watch it, write it again"...SKIP
Test [798/1614] crypto.pbkdf2.test "std-native-Debug-bare-multiRFC 6070 16,777,216 iterations"...SKIP
Test [801/1614] crypto.pbkdf2.test "std-native-Debug-bare-multiVery large dkLen"...SKIP
Test [851/1614] crypto.25519.x25519.test "std-native-Debug-bare-multix25519 rfc7748 1,000 iterations"...SKIP
Test [852/1614] crypto.25519.x25519.test "std-native-Debug-bare-multix25519 rfc7748 1,000,000 iterations"...SKIP
Test [855/1614] event.channel.test "std-native-Debug-bare-multistd.event.Channel"...SKIP
Test [856/1614] event.channel.test "std-native-Debug-bare-multistd.event.Channel wraparound"...SKIP
Test [857/1614] event.future.test "std-native-Debug-bare-multistd.event.Future"...SKIP
Test [858/1614] event.group.test "std-native-Debug-bare-multistd.event.Group"...SKIP
Test [860/1614] event.lock.test "std-native-Debug-bare-multistd.event.Lock"...SKIP
Test [861/1614] event.rwlock.test "std-native-Debug-bare-multistd.event.RwLock"...SKIP
Test [862/1614] event.wait_group.test "std-native-Debug-bare-multibasic WaitGroup usage"...SKIP
Test [1215/1614] net.test.test "std-native-Debug-bare-multiinvalid but parseable IPv6 scope ids"...SKIP
Test [1219/1614] net.test.test "std-native-Debug-bare-multilisten on a port, send bytes, receive bytes"...SKIP
Test [1220/1614] net.test.test "std-native-Debug-bare-multilisten on ipv4 try connect on ipv6 then ipv4"...SKIP
1559 passed; 55 skipped.
3 errors were logged.
1 tests leaked memory.
error: the following test command failed with exit code 1:
D:\git\zig\zig-cache\o\a88e7ec65ce2a4e4aa9b8c6d3960e7ed\test.exe

```

**Comment by marler8997** [comment]
The issue I see with this change is that we're making it harder to write correct Zig code.  Consider this example:
```zig
const msg = "...some reason...";
std.log.error("this thing failed because {}", .{msg});
```

With this change we'll see something like this:
```
this thing failed because u8 { 39, 39, 39, 58, 20, ...}
```

Because Zig aims to handle all error cases correctly, error messages are an important part of any codebase.  The problem here is now we're making it easy to write an incorrect error message and it will silently fail until someone sees that error message and files a bug report.

That being said, I agree the status quo is also not ideal.  Having special casing for these types adds complexity.

What about a modification to this change? What if we always required a specifier for all array types? If we did that, then this problem is now caught at compile time:
```Zig
const msg = "...some reason...";
std.log.error("this thing failed because {}", .{msg});
```
```
error: cannot print array type `msg` without a specifer (i.e. {s} or {x})
```

**Comment by LemonBoy** [comment]
> What about a modification to this change? What if we always required a specifier for all array types? If we did that, then this problem is now caught at compile time:

I'm @LemonBoy and I approve this message :+1:

**Comment by Tetralux** [comment]
I would also mention `{d}` (for decimal, IIRC) in the message, if you're gonna mention `{s}` and `{x}`.

**Comment by marler8997** [comment]
Yeah I tried building debug already and I remember this error:
```
cl : command line error D8016: '/O2' and '/RTC1' command-line options are incompatible [D:\a\1\s\build\opt_c_util.vcxproj]
```
You're feeling the effects of Windows getting no love.

**Comment by marler8997** [comment]
FYI, even if you remove the `/O2` flag, I don't think it works unless you link against the Debug llvm binaries.

**Comment by LemonBoy** [comment]
I've basically reverted every change done to fmt and left only the `{}` to {s}` changes but the windows CI is still failing.

If anyone is able to reproduce this problem please get in touch, I've ran out of ideas.

**Comment by FireFox317** [comment]
Will try to debug on Windows now.

**Comment by FireFox317** [comment]
This is probably caused by the Windows CI going OOM. It basically happends when the the compiler is compiling all the tests for `test-std`. Everything ends up in a single compile unit and the CI does not have enough memory to compile all this. We have had this problem before and I think it happend at the exact same place.

On my machine it maxed out at 7GB of memory to compile the tests...

All std tests pass locally btw!

**Comment by LemonBoy** [comment]
> This is probably caused by the Windows CI going OOM. It basically happends when the the compiler is compiling all the tests for test-std. Everything ends up in a single compile unit and the CI does not have enough memory to compile all this. We have had this problem before and I think it happend at the exact same place.

The question is, does this PR increase the RAM consumption? If so, how?

**Comment by FireFox317** [comment]
> The question is, does this PR increase the RAM consumption? If so, how?

Good question, I have no idea, but apparently it does.

**Comment by andrewrk** [comment]
I love this and want it, let's coordinate on how to land this since it is destined to conflict with everything

**Comment by LemonBoy** [comment]
> I love this and want it, let's coordinate on how to land this since it is destined to conflict with everything

I've rebased the whole patch series, I don't expect many conflicts with other PRs as the changes are pretty much self-contained.

> the amount of RAM we get on Windows CI runs is not consistent. so it doesn't necessarily mean this PR causes more ram consumption.

*shrug*, then I wonder why we always hit the OOM condition with this PR...
We have the same problem in `master`and #7331.

**Pull request closed without merging**